docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,188 @@
+
+As a CAPZ user I want to take advantage of the latest and greatest features in Azure only available in the newest Azure API versions in my clusters managed by CAPZ. This would allow me to adopt CAPZ for a wider range of use cases and continue evolving my CAPZ clusters alongside the Azure API.
+
+#### Story 2 - Developer experience
+
+As a CAPZ developer I would like to only need to interact with the Kubernetes API server using Custom Resource Definitions representing Azure resources rather than having to deal with the impedance of interacting with the Azure APIs. This would enable me to focus more on the core value of CAPZ, transforming Cluster API resource definitions to Azure resource definitions.
+
+### Implementation Details/Notes/Constraints
+
+- ASO needs to be installed on the management cluster
+  - Multi- or single-tenant?
+  - Install in capz-controller-manager `func main()`
+- GET requests are not as straightforward in the ASO model because BYO resources
+  may not initially exist in ASO
+- Current service pattern in CAPZ works for Azure services that are not
+  represented in ASO (resource health)
+- Each release of ASO can expose several Azure API versions of each resource,
+  including GA and preview API versions.
+- Must ensure CAPZ only operates against the ASO resources it manages, not ones managed by the user.
+
+### Security Model
+
+ASO requires credentials to interact with Azure which are documented here: https://azure.github.io/azure-service-operator/introduction/authentication/. Specifically, ASO needs the values of `$AZURE_CLIENT_ID`, `$AZURE_CLIENT_SECRET`, `$AZURE_SUBSCRIPTION_ID`, and `$AZURE_TENANT_ID` for Service Principal authentication, which are already set in the capz-controller-manager container's environment variables. Passing them along as chart values when installing ASO with Helm doesn't introduce any significant additional security risks.
+
+Consumers of ASO's APIs (such as CAPZ) also need Kubernetes RBAC permissions to interact with ASO resources. For example, this ClusterRole snippet is required by CAPZ to manage resource groups with ASO:
+```yaml
+- apiGroups:
+  - resources.azure.com
+  resources:
+  - resourcegroups
+  - resourcegroups/status
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
+```
+The above permissions may overlap with ASO resources already installed by a user not intended to be managed by CAPZ.
+
+### Risks and Mitigations
+
+- Introduction of functional regressions
+- Introduction of performance regressions, where operations against workload
+  clusters may take longer to complete
+- Introduction of scalability regressions, where management clusters may exhaust
+  Azure or Kubernetes API limits with fewer or smaller workload clusters being managed.
+  - Management cluster will have to manage many more Kubernetes resources per
+    workload cluster
+- Because ASO has not yet been proven as a mission-critical interface to Azure
```

Comment:
Well-phrased. I agree with this as a risk.

I think it makes a good bit of sense to make a shared bet. As you called out, ASO is solving the "2. Interfacing with the Azure platform to manage creating, updating, and deleting that infrastructure" problem, so it should end up reducing the work CAPZ has to do on that stuff, but this is a risk as obviously the Azure Go SDK has much broader adoption and is more mature (GA) than ASO is currently.